### DISCUSSION TITLE
What is the correct way to create a custom port so that the vcpkg can use it?

### DISCUSSION THREAD

**Discussion by matusnovak** [microsoft/vcpkg#27742] [Q&A] [answered]
I am trying to create a custom library (uses CMake) and make it available to use by vcpkg. I am following this tutorial: https://devblogs.microsoft.com/cppblog/registries-bring-your-own-libraries-to-vcpkg/

My library (let's call it `Example` target) CMakeLists.txt contains the following:

```
install(TARGETS Example
        EXPORT ExampleConfig
        RUNTIME DESTINATION ${CMAKE_INSTALL_BINDIR}
        LIBRARY DESTINATION lib
        ARCHIVE DESTINATION lib
        )

install(DIRECTORY "${CMAKE_CURRENT_LIST_DIR}/src/" DESTINATION include FILES_MATCHING PATTERN "*.hpp")

export(TARGETS
        Example
        NAMESPACE Example::
        FILE "${CMAKE_CURRENT_BINARY_DIR}/ExampleConfig.cmake"
        )
install(EXPORT
        ExampleConfig
        DESTINATION "share/Example"
        NAMESPACE Example::
        )
```

And my custom port file contains the following:

```
vcpkg_from_github(
  OUT_SOURCE_PATH SOURCE_PATH
  REPO ...
  REF ...
  SHA512 ...
  HEAD_REF master
)

vcpkg_configure_cmake(
  SOURCE_PATH "${SOURCE_PATH}"
  PREFER_NINJA
)
vcpkg_install_cmake()

file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug/include") # Not sure why this is needed
```

This port file is located on my system. I try to install the port with:

```
vcpkg install example --overlay-ports=vcpkg-registry/ports/example
```

The port is recognized, downloads the report, builds a release and debug. So far everything is good.

However, it prints out the following:

```
find_package(cmake CONFIG REQUIRED)
target_link_libraries(main PRIVATE Example::example)
```

**Notice the `cmake` in `find_package`. That is auto generated. Why is that? Why doesn't it contain the correct name `example`?**

Moreover, the config files are installed in the following folder: `/path/to/vcpkg/installed/x64-linux/debug/share/Example/ExampleConfig.cmake`

**Why does it go into the `debug` folder? I would expect it go to the `vcpkg/installed/x64-linux/share/Example` folder, where the rest of the installed libraries are. Why is that?**

If there is any good tutorial on how to create custom ports, please do share. The current state of documentation on how to correctly do this terrible.

**Comment by Neumann-A** [comment, marked as answer]
Consider using `vcpkg-cmake` and `vcpkg-cmake-config-fixup `as dependencies and update to the newer port functions. Also you are missing a call to `vcpkg_cmake_config_fixup/vcpkg_fixup_cmake_targets`

**Reply by matusnovak** [reply]
I can confirm that adding `vcpkg_cmake_config_fixup()` and `vcpkg_fixup_cmake_targets()` after `vcpkg_configure_cmake(...)` solved it. Alongside with:
```
file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug/include")
file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug/share")
```

Thank you.
